versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java

```diff
@@ -114,17 +118,55 @@ public Hash hashOnReference(NamedRef namedReference, Optional<Hash> hashOnRefere
   }
 
   @Override
-  public Stream<WithHash<NamedRef>> namedRefs() {
+  public Stream<NamedReference<ByteString>> namedRefs(NamedReferencesParams params) {
+    Preconditions.checkNotNull(params, "Parameter for NamedReferencesParams must not be null");
+    Preconditions.checkArgument(
+        params.isRetrieveTags() || params.isRetrieveBranches(),
+        "Must retrieve branches or tags or both.");
+
     GlobalStatePointer pointer = fetchGlobalPointer(NON_TRANSACTIONAL_OPERATION_CONTEXT);
     if (pointer == null) {
       return Stream.empty();
     }
-    return pointer.getNamedReferencesMap().entrySet().stream()
-        .map(
-            r ->
-                WithHash.of(
-                    Hash.of(r.getValue().getHash()),
-                    toNamedRef(r.getValue().getType(), r.getKey())));
+
+    Hash defaultBranchHead = defaultBranchHeadForNamedRefs(params, pointer);
+
+    Stream<NamedReference<ByteString>> refs =
+        pointer.getNamedReferencesMap().entrySet().stream()
+            .filter(
+                r ->
+                    (params.isRetrieveBranches()
+                            && r.getValue().getType() == RefPointer.Type.Branch)
+                        || (params.isRetrieveTags()
+                            && r.getValue().getType() == RefPointer.Type.Tag))
+            .map(
+                r ->
+                    NamedReference.<ByteString>builder()
+                        .namedRef(toNamedRef(r.getValue().getType(), r.getKey()))
+                        .hash(Hash.of(r.getValue().getHash()))
+                        .build());
+
+    NonTransactionalOperationContext ctx = NON_TRANSACTIONAL_OPERATION_CONTEXT;
+    refs = namedRefsWithDefaultBranchRelatedInfo(ctx, refs, params, defaultBranchHead);
+
+    if (params.isRetrieveHead()) {
+      refs = refs.map(r -> namedReferenceWithCommitMeta(NON_TRANSACTIONAL_OPERATION_CONTEXT, r));
+    }
+
+    return refs;
+  }
+
+  private Hash defaultBranchHeadForNamedRefs(
```

Comment:
maybe let's mark this as `@Nullable`

Reply:
Can do in a separate PR (for all interesting places), but would like to keep it out of this one